Desktop_Interface/isobuffer.cpp

```diff
@@ -2,391 +2,364 @@
 #include "isodriver.h"
 #include "uartstyledecoder.h"
 
-isoBuffer::isoBuffer(QWidget *parent, int bufferLen, isoDriver *caller, unsigned char channel_value) : QWidget(parent)
+isoBuffer::isoBuffer(QWidget* parent, int bufferLen, isoDriver* caller, unsigned char channel_value)
+	: QWidget(parent)
 {
-    buffer = (short *) calloc(bufferLen*2, sizeof(short));
-    bufferEnd = bufferLen-1;
-    samplesPerSecond = (double) bufferLen/(double)21;
-    samplesPerSecond = samplesPerSecond/375*VALID_DATA_PER_375;
-    sampleRate_bit = samplesPerSecond * 8;
-    virtualParent = caller;
-    channel = channel_value;
+	buffer = (short*) calloc(bufferLen*2, sizeof(short));
+	bufferEnd = bufferLen-1;
+	samplesPerSecond = (double) bufferLen/(double)21;
+	samplesPerSecond = samplesPerSecond/375*VALID_DATA_PER_375;
+	sampleRate_bit = samplesPerSecond * 8;
+	virtualParent = caller;
+	channel = channel_value;
 }
+
 void isoBuffer::openFile(QString newFile)
 {
-    if (fptr != NULL){
-        fclose(fptr);
-    }
-    if (newFile.isEmpty()){
-        fptr = NULL;
-    }
-    else {
-        QByteArray temp = newFile.toLatin1();
-        char *fileName = temp.data();
-        fptr = fopen(fileName, "w");
-        if (fptr == NULL) qFatal("Null fptr in isoBuffer::openFile");
-        qDebug() << "opening file" << fileName;
-        qDebug() << "fptr = " << fptr;
-    }
+	if (fptr != NULL)
+	{
+		fclose(fptr);
+	}
+	if (newFile.isEmpty())
+	{
+		fptr = NULL;
+	}
+	else
+	{
+		QByteArray temp = newFile.toLatin1();
+		char* fileName = temp.data();
+		fptr = fopen(fileName, "w");
+		if (fptr == NULL) qFatal("Null fptr in isoBuffer::openFile");
+		qDebug() << "opening file" << fileName;
+		qDebug() << "fptr = " << fptr;
+	}
+}
+
+void isoBuffer::insertIntoBuffer(short item)
+{
+	buffer[back] = item;
+	back++;
+	insertedCount++;
+	if (insertedCount > bufferEnd)
+	{
+		insertedCount = bufferEnd+1;
+	}
+	if (back > bufferEnd)
+	{
+		back = 0;
+	}
+}
+
+short isoBuffer::bufferAt(int idx) const
+{
+	return buffer[back - idx];
+}
+
+bool isoBuffer::maybeOutputSampleToFile(double convertedSample)
+{
+	/*
+	 * This function adds a sample to an accumulator and bumps the sample count
```

Comment:
The accumulator is used to take sample averages.  For example, in the case where the user wants to output one data sample to a file per second but the scope reads at 750ksps, this will average 750,000 samples and then write the average to the file.
However, there can be a long time and multiple functions calls before it actually gets that many samples, hence the need for the long-term accumulator.[